### ISSUE TITLE
Vue 3

### ISSUE THREAD

**Issue by pascalgrimaud** [jhipster/generator-jhipster#12558]
##### **Overview of the feature request**

Vue 3 has been released ! 
It would be so cool if we can have it for JHipster v7

##### **Related issues or PR**

-   [x] Checking this box is mandatory (this is just to show you read everything)

**Comment by pascalgrimaud** [comment]
cc @deepu105 @jdubois 
I'm adding a big bounty on this, as I think there is a lot of work to achieve this, but it would be really so cool to have it !

**Comment by atomfrede** [comment]
Let's see how fast then 3rd party components are getting compatible. They have postponed the migration utilities for now as far as I know (but not sure how much it would help in our case).

**Comment by pascalgrimaud** [comment]
@atomfrede : do you have more information ? I thought it would be compatible with Vue 3

**Comment by atomfrede** [comment]
Did not look in detail yet. I also guess we can upgrade without too much hassle. We have few 3rd party libraries and they are widely used. I try to have a look next week to get a better understanding what would be needed.

**Comment by pascalgrimaud** [comment]
I'd love to have this for JHipster v7
Don't hesitate to ask help, and if needed, we can add separate bounties for people who want to help here

**Comment by atomfrede** [comment]
At least bootstrap-vue is not yet compatible. They plan to have a compatible release "short" after vue 3 release. Will follow closely. https://github.com/bootstrap-vue/bootstrap-vue/issues/5196#issuecomment-668837891

**Comment by atomfrede** [comment]
For i18n there is already a beta version which we could try and use https://github.com/intlify/vue-i18n-next.

Both [veeValidate](https://logaretm.github.io/vee-validate) and [Vuelidate](https://vuelidate.js.org/) are not yet compatible out of box with vue3. But using the [composition api](https://itnext.io/compose-validation-logics-with-vue-3-0-and-vuelidate-20416d734295) it might be possible. I will try to finish the (finally) the [eslint migration ](https://github.com/jhipster/jhipster-vuejs/pull/601) in the main generator and can try vue3 afterwards.

**Comment by chenhui7373** [comment]
vue3 compatible: https://2x.antdv.com/docs/vue/introduce-cn/

**Comment by mrts** [comment]
@atomfrede note that a ["migration build"](https://v3.vuejs.org/guide/migration/introduction.html#breaking-changes) is in the works with Vue 2 compatible behavior (and runtime warnings of incompatible usage). So perhaps an interim step would be to implement minimal changes to use the migration build even if this is not fully optimal and then follow up with a rewrite to composition-API eventually.

Note that there is also [Vite](https://github.com/vitejs/vite) for Vue 3 which makes frontend development lightning fast.

**Comment by atomfrede** [comment]
As far as I understand the migration build is not yet available?

`We are still working on a dedicated Migration Build of Vue 3 with Vue 2 compatible behavior and runtime warnings of incompatible usage. If you are planning to migrate a non-trivial Vue 2 app, we strongly recommend waiting for the Migration Build for a smoother experience.`

**Comment by mrts** [comment]
> As far as I understand the migration build is not yet available?

Yes, it is not yet available. According to [Vue 3 roadmap](https://github.com/vuejs/vue/projects/6), it is planned for Q4 2020.

**Comment by pascalgrimaud** [comment]
simple comment to keep this opened:)

**Comment by gmarziou** [comment]
The keep-alive comment :)

**Comment by atomfrede** [comment]
Keep it open.

**Comment by mraible** [comment]
Should we add a bounty for motivation?

**Comment by atomfrede** [comment]
We are depending on vue bootstrap compatibility I think. Maybe we can start by updating to vue i18n and and replace the current client side validation library with compatibility one as a starting point maybe.

**Comment by atomfrede** [comment]
I would suggest the following to make the eventual v3 migration a bit easier:

* replace [vuelidate](https://vuelidate.js.org/) with [veeValidate](https://vee-validate.logaretm.com/v3/) as vuelidate is not well maintained anymore and I don't see it will be compatible with vue3 anytime
* Update to https://github.com/intlify/vue-i18n-next which is the successor of the current i19n solution but compatible with vue2 and 3

Having this we have already achieved large parts to be compatible with vue 3. The remaining part is vue bootstrap. Without that I don't think we can do the switch easily. So we have to wait for a compatible release or the migration build (which is still work on progress) and try it with the current vue boostrap release. 
I don't think we have the capacity to just use plain bootstrap and write/maintain our own components for e.g. modals etc.

**Comment by michaelzangl** [comment]
This is the vue3 for bootstrap-vue issue:
https://github.com/bootstrap-vue/bootstrap-vue/issues/5196

**Comment by atomfrede** [comment]
It might take some more time. Estimation seems to be a few weeks for alpha/beta to be ready, stable in Q2 if everything goes well. https://github.com/bootstrap-vue/bootstrap-vue/issues/5196#issuecomment-770883757

**Comment by atomfrede** [comment]
Please keep it open we need to wait on 3rd party dependencies

**Comment by atomfrede** [comment]
Keep it open.

**Comment by atomfrede** [comment]
Keep it open

**Comment by atomfrede** [comment]
Vue 3.1.0 has been released as a first beta Version. This contains the compat build. We could give it a try https://github.com/vuejs/vue-next/tree/master/packages/vue-compat

**Comment by mrts** [comment]
Just a quick heads-up that there is some discussion regarding using the compat build in VueBootstrap issue tracker here:  https://github.com/bootstrap-vue/bootstrap-vue/issues/5196. In short, the compat build works with simpler components, but more complex components break. So it is not ready for prime time yet.

**Comment by atomfrede** [comment]
Keep it open

**Comment by shershen08** [comment]
in addition to a summary https://github.com/jhipster/generator-jhipster/issues/12558#issuecomment-753474557 by @atomfrede  

As I understood https://github.com/bootstrap-vue/bootstrap-vue/issues/5196 doesn't have candidate MR that's being actively worked on, it's more a collection of problems and links how people approach Vue2->Vue3 upgrade in BootstrapVue community.
There's also this rewrite from scratch Vue3+BS5 - https://github.com/cdmoro/bootstrap-vue-3, but it's missing a lot of functionality still.

In the current Vue generator it's 65 components in total, some are relatively simple:
<b-form 
<b-form-input
<b-button 
<b-link
<b-alert 
<b-progress-bar
<b-nav-item 
<b-navbar-nav 
<b-collapse 
<b-dropdown-item 

but some are more complex:
<b-pagination 
<b-modal 

we could think of either replacing them with pure Bootstrap5 or separate libraries

**Comment by atomfrede** [comment]
Keep it open

**Comment by chen1i** [comment]
keep it open. looking forward to having it real...

**Comment by mraible** [comment]
Keep it open! Are there any volunteers that can look into this for $500? If not, I can look into it as I've done a migration in the past.

**Comment by rs4231199** [comment]
@mraible

> Keep it open! Are there any volunteers that can look into this for $500? 

 I will do it.

**Comment by gmarziou** [comment]
Using PrimeVue would make sense if JHipster Angular would use PrimeNG which I would love but I doubt it will happen.

**Comment by mrts** [comment]
@gmarziou, considering that the big three are covered (PrimeNG, PrimeReact, PrimeVue), PrimeFaces looks really attractive for JHipster. What are the downsides of using it?

**Comment by mraible** [comment]
I think the downside is the maintainers know Bootstrap better. It might be
a good idea to create a blueprint or module for this.

**Comment by mrts** [comment]
By the way, I would strongly suggest that we use `<script setup>` that was introduced in Vue 3.2. It is so good that it hurts - you just write variables and functions in the `<script setup>` block and they become available in the template scope automatically and become reactive when you use `ref`.

Here's a minimal example from the [release blog post](https://blog.vuejs.org/posts/vue-3.2.html):
```vue
<script setup>
import { ref } from 'vue'

const color = ref('red')
</script>

<template>
  <button @click="color = color === 'red' ? 'green' : 'red'">
    Color is: {{ color }}
  </button>
</template>

<style scoped>
button {
  color: v-bind(color);
}
</style>
```

And [here's a simple to-do list component](https://github.com/marcushellberg/vue-todo/blob/master/src/components/TodoView.vue) that loads data from and saves it to a REST API.

**Comment by mshima** [comment]
Having same toolkit for every client framework is hand for maintenance.
They share the same translation process, same rtl support, same webpack plugins for specific things like notification, microfrontend, browser sync.
Hopefully this new development finally will get us there https://github.com/bootstrap-vue/bootstrap-vue/pull/6845.

**Comment by mraible** [comment]
Keep it open.

**Comment by mraible** [comment]
$500 bounty! Maybe we should split this one into smaller tasks to inspire contributions?

**Comment by Soromeister** [comment]
Still, keep it open.

**Comment by pascalgrimaud** [comment]
I wonder if someone wants to start the work, otherwise, I'll close the ticket as lack of interest, because this ticket has been opened since 2 years

For your information, the support of Vue3 is already in JHipster Lite (since 6 months maybe), so I won't work on generator-jhipster to add this.

**Comment by mrts** [comment]
@pascalgrimaud, which component library did you use in JHipster Lite? The issue is currently blocking behind Bootstrap Vue Vue 3 support. The developer who deals with adding Vue 3 support lives in Ukraine and struggling to find balance between war-related activities, work and his Bootstrap Vue work.

See current status update here: https://github.com/bootstrap-vue/bootstrap-vue/pull/6905#issuecomment-1149083451

**Comment by pascalgrimaud** [comment]
@mrts : in JHLite, we don't use bootstrap-vue, but we use bootstrap directly and we plan to do the same for the generated part (which is just a home page today):

https://github.com/jhipster/jhipster-lite/blob/main/package.json#L33-L40

So we don't rely to bootstrap-vue and we don't need to wait the work there

**Comment by pascalgrimaud** [comment]
in JH Lite, we use Hexagonal architecture, so we don't care too much about the infrastructure. The logic code is inside the domain, and this part is the most important. The secondary adapter do not use bootstrap, only axios
So bootstrap is related to the primary adapter only and migrate this part to another framework is easy

**Comment by mrts** [comment]
Alright, thanks for the explanation!

**Comment by pascalgrimaud** [comment]
Additional information @mrts : by using this architecture, we manage to generate project with 100% coverage, 0 code smell, 0 duplication, etc. Best quality as possible. I already have several projects in production for my customer, using this little project :-p
And the engine to generate project uses the same kind of architecture too. See the result (22k lines of code):

![image](https://user-images.githubusercontent.com/9156882/174298655-8bf672b3-d38b-4fb4-ba14-51e7b12cd139.png)

**Comment by pascalgrimaud** [comment]
Anyway, sorry for this out of subject: if no body wants to start the work on Vue3, I'll close this ticket. We can open smaller tickets to ease the contribution and help the contributors

**Comment by mrts** [comment]
I would prefer to keep it open, although I'm not a JHipster team member and cannot influence decision-making. We are waiting for  Bootstrap Vue Vue 3 support that should arrive eventually and it is possible to start the work then - so it is not a question of no interest but rather about blocking on a dependency.

**Comment by mraible** [comment]
I agree it would be nice to have Vue 3 support in the main JHipster generator.

**Comment by michaelzangl** [comment]
> @pascalgrimaud, are you aware that using Bootstrap directly might require extensive work with Bootstrap JavaScript code? See [bootstrap-vue/bootstrap-vue#5196 (comment)](https://github.com/bootstrap-vue/bootstrap-vue/issues/5196#issuecomment-893312245).

Just don't include that code. I work with Vue3+Bootstrap for almost 2 years now. Most of the bootstrap JS code is just for e.g. dismissing alerts. This mechanism can easily be re-build in Vue. I only use Bootstrap for Styling.

**Comment by Giwayume** [comment]
@michaelzangl 

The referenced library does a whole lot more than just dismissing alerts, it has a feature-complete WAI ARIA A11y integration, which simply pretty much all other libraries don't have. It is uniquely valuable for that facet alone.

**Comment by Giwayume** [comment]
It's easy to ignore that when you have no experience in building websites in an accessible way.

**Comment by mraible** [comment]
Is there anything currently blocking this upgrade? I ask because I've added to my list of issues to work on at the JavaOne Hackergarten next week (October 18). https://bit.ly/javaone22-jhipster

**Comment by mrts** [comment]
Bootstrap Vue Vue 3 support is unfortunately still not ready. @Giwayume's point on the usefulness of Bootstrap Vue still stands in my opinion (as opposed to @michaelzangl's view of using plain Bootstrap).

See the current status update here: https://github.com/bootstrap-vue/bootstrap-vue/pull/6905#issuecomment-1272108708.

**Comment by mshima** [comment]
I am against using pure bootstrap too. This will come with too much maintenance burden.
Since we are light user and official BootstrapVue support for bootstrap 5 looks far, we may look for alternatives.
An alternative is to use [bootstrap vue 3](https://github.com/cdmoro/bootstrap-vue-3). Which will bring bootstrap 5. And a maintainer is seeking for [cooperation](https://github.com/bootstrap-vue/bootstrap-vue/pull/6905#issuecomment-1272095733)

**Comment by mshima** [comment]
Since we are at v8 cycle, I suggest to migrate to vite which is the recommended now.
The problematic migration may be microfrontend with module-federation. I can try to help with that.
Migration info is available at https://bootstrap-vue.org/vue3.

**Comment by mshima** [comment]
I will start modularizing vue generator and migrating if nobody started.

**Comment by mshima** [comment]
I will finish the modularizing, but we will need to drop some libs (decorator, class) before migrating.
So I will not implement the migrating for now.

**Comment by staaph** [comment]
Any progress?

**Comment by mshima** [comment]
I've managed to convert a project to vue3 and compat mode.
Unfortunately there is TOO much work.
I will open a new bug to covert vue to [composition api](https://vuejs.org/guide/extras/composition-api-faq.html#what-is-composition-api).

The steps would be:
- vue-cli https://github.com/jhipster/generator-jhipster/pull/15932
- convert to composition api
- convert to vue 3 (deprecations and rework).

**Comment by mshima** [comment]
@DanielFran bug bounty claimed https://opencollective.com/generator-jhipster/expenses/140219

**Comment by DanielFran** [comment]
@mshima approved
